### PR TITLE
Adding an example on AdaMatch

### PR DESCRIPTION
**Pull request by sayakpaul**
This example introduces important concepts like semi-supervised learning and domain adaptation. At the same time, it also implements [AdaMatch](https://arxiv.org/abs/2106.04732), which currently achieves SOTA on unsupervised domain adaptation. 

Accompanying Colab Notebook is available [here](https://colab.research.google.com/drive/1OLXyHAzgU15K-JBm7YWk3Hvat60lZJJW?usp=sharing).

### CODE REVIEW

File: examples/vision/adamatch.py [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+final_source_ds = tf.data.Dataset.zip((source_ds_w, source_ds_s))
+
+target_ds_w = create_individual_ds(svhn_train, weak_augment, source=False)
+target_ds_s = create_individual_ds(svhn_train, strong_augment, source=False)
+final_target_ds = tf.data.Dataset.zip((target_ds_w, target_ds_s))
+
+"""
+Here's how the image batches look like:
+
+![](https://i.ibb.co/FDVkbWL/Data-Viz.png)
```
Comment (fchollet) [inline examples/vision/adamatch.py:188]: The link doesn't work for me. If possible, use code-based visualization, like Matplotlib.

File: examples/vision/adamatch.py [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+    # First compute the losses between original source labels and
+    # predictions made on the weakly and strongly augmented versions
+    # of the same images.
+    w_loss = loss_func(source_labels, logits_source_w)
+    s_loss = loss_func(source_labels, logits_source_s)
+    return w_loss + s_loss
+
+
+def compute_loss_target(target_pseudo_labels_w, logits_target_s, mask):
+    loss_func = keras.losses.categorical_crossentropy
```
Comment (fchollet) [inline examples/vision/adamatch.py:207]: Why is this one not from logits if the other one is? This is a bit confusing

File: examples/vision/adamatch.py [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+
+
+"""
+Rather than using a fixed scalar quantity, a varying scalar is used in AdaMatch. It
+denotes the weight of the loss contibuted by the target samples.
+"""
+
+
+def mu_schedule():
+    global T
```
Comment (fchollet) [inline examples/vision/adamatch.py:227]: The use of a global here is an anti-pattern. You should make this function into a model method, and make T (rename it to its full name) into a model property.

File: examples/vision/adamatch.py [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+
+svhn_train, svhn_test = tfds.load(
+    "svhn_cropped", split=["train", "test"], as_supervised=True
+)
+
+"""
+## Constants and hyperparameters
+"""
+
+TF_PI = tf.constant(np.pi)
```
Comment (fchollet) [inline examples/vision/adamatch.py:97]: You can probably just inline it

File: examples/vision/adamatch.py [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+"""
+## Instantiate a Wide-ResNet-28-2
+
+The authors use a [WideResNet-28-2](https://arxiv.org/abs/1605.07146) for the dataset
+pairs we are using in this example. Note that the following model has a scaling layer
+inside it that scales the pixel values to [0, 1]. 
+"""
+
+"""shell
+wget -O wide_resnet.py -q https://git.io/Jnzzj
```
Comment (fchollet) [inline examples/vision/adamatch.py:393]: We should definitely not be downloading arbitrary code to execute. Try to in-line the code in a function. If not possible, replace it with a generic Keras Application.

File: examples/vision/adamatch.py [hunk 6]
```diff
@@ -0,0 +1,453 @@
...
+        ds = ds.map(lambda x, y: (aug_func(x), y), num_parallel_calls=AUTO)
+    else:
+        ds = ds.map(lambda x, y: (aug_func(x, False), y), num_parallel_calls=AUTO)
+
+    ds = ds.batch(batch_size).prefetch(AUTO)
+    return ds
+
+
+"""
+`_w` and `_s` suffixes denote weak and strong respectively.
```
Comment (fchollet) [inline examples/vision/adamatch.py:173]: Prefer using fully spelled-out variable names for better readability.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR -- very cool new example!

File: examples/vision/adamatch.py (hunk 1 above)
Comment (sayakpaul) [reply examples/vision/adamatch.py:188]: I intentionally skipped that part to reduce the lines of code. Could you copy-paste the link and open it separately? Because it works for me even in Incognito mode.

File: examples/vision/adamatch.py (hunk 2 above)
Comment (sayakpaul) [reply examples/vision/adamatch.py:207]: I am using `from_logits=True` here:

```
target_loss = loss_func(target_pseudo_labels_w, logits_target_s, from_logits=True)
```

Since the reduction of this loss function is a bit more involved I used the `categorical_crossentropy` function rather than the class.

File: examples/vision/adamatch.py (hunk 3 above)
Comment (sayakpaul) [reply examples/vision/adamatch.py:227]: Sure. The value of `T` represents the current training step. I was wondering how to best achieve that and `global` is what I could think about. Could you expand on this a bit more?

File: examples/vision/adamatch.py (hunk 5 above)
Comment (sayakpaul) [reply examples/vision/adamatch.py:393]: Sure. I will inline the function. I was hesitant to do that to reduce the lines of code.

File: examples/vision/adamatch.py (hunk 6 above)
Comment (sayakpaul) [reply examples/vision/adamatch.py:173]: It was creating very big variable names and was breaking the readability in some cases. This is why I used shorthand notations.

File: examples/vision/adamatch.py (hunk 2 above)
Comment (fchollet) [reply examples/vision/adamatch.py:215]: Then you should do `loss_func = keras.losses.CategoricalCrossentropy(from_logits=True, reduction='none')`
then `target_loss  = loss_func(target_pseudo_labels_w, logits_target_s)`
to be consistent with the source loss

File: examples/vision/adamatch.py (hunk 1 above)
Comment (fchollet) [reply examples/vision/adamatch.py:196]: Try using imgur.com.

File: examples/vision/adamatch.py (hunk 3 above)
Comment (fchollet) [reply examples/vision/adamatch.py]: I think the current implementation isn't doing what you think it's doing. `T` is a Python variable, so when `train_step` gets compiled it will capture the value of the variable at the time of compilation. It will never get incremented. Here's what you should do:

```python
class AdaMatch(keras.Model):
  def __init__(self, total_steps, ...):
    super(AdaMatch, self).__init__()
    self.total_steps = total_steps
    self.current_step = tf.Variable(0, dtype='int64')
    ...

  def compute_mu(self):
    pi = tf.constant(np.pi, dtype='float32')
    step = tf.cast(self.current_step, dtype='float32')
    return 0.5 - tf.cos(tf.math.minimum(pi, (2 * pi * step) / self.total_steps)) / 2)

  def train_step(self, data):
    ...
    t = self.compute_mu()
    self.current_step.assign_add(1)
    ...
```

Does this make sense?

Comment (sayakpaul) [reply examples/vision/adamatch.py]: Totally. Thank you.

**Comment by sayakpaul** [comment]
@fchollet I have addressed all the feedback. Let me know if I should proceed to add the generated files.

File: examples/vision/adamatch.py [hunk 7]
```diff
@@ -0,0 +1,606 @@
...
+"""
+
+RESIZE_TO = 32
+
+SOURCE_BATCH_SIZE = 64
+TARGET_BATCH_SIZE = 3 * SOURCE_BATCH_SIZE  # Reference: Section 3.2
+EPOCHS = 10
+STEPS_PER_EPOCH = len(mnist_x_train) // SOURCE_BATCH_SIZE
+TOTAL_STEPS = EPOCHS * STEPS_PER_EPOCH
+CURRENT_STEP = 0
```
Comment (fchollet) [inline examples/vision/adamatch.py:111]: You don't need this any more

File: examples/vision/adamatch.py [hunk 8]
```diff
@@ -0,0 +1,606 @@
...
+"""
+
+
+class AdaMatch(keras.Model):
+    def __init__(self, model, tau=0.9):
+        super(AdaMatch, self).__init__()
+        self.model = model
+        self.tau = tau  # Denotes the confidence threshold
+        self.loss_tracker = tf.keras.metrics.Mean(name="loss")
+        self.total_steps = TOTAL_STEPS
```
Comment (fchollet) [inline examples/vision/adamatch.py:265]: Make this a constructor argument, then do `model = AdaMatch(total_steps= TOTAL_STEPS)`. This improves code reusability. The internal implementation of a model should not depend on outer global constants.

**Review by fchollet — COMMENTED** [review]
Thanks for the update! I made some copyedits, please pull them first.

File: examples/vision/adamatch.py [hunk 9]
```diff
@@ -0,0 +1,604 @@
...
+The authors use a [WideResNet-28-2](https://arxiv.org/abs/1605.07146) for the dataset
+pairs we are using in this example. Most of the following code has been referred from
+[this script](https://github.com/asmith26/wide_resnets_keras/blob/master/main.py). Note
+that the following model has a scaling layer inside it that scales the pixel values to
+[0, 1].
+"""
+
+
+def wide_basic(n_input_plane, n_output_plane, stride):
+    def f(net):
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply examples/vision/adamatch.py]: Sorry about this. I have made the necessary change.

File: examples/vision/adamatch.py (hunk 7 above)
Comment (sayakpaul) [reply examples/vision/adamatch.py]: Done.

**Comment by sayakpaul** [comment]
@fchollet applied your recent suggestions. Let me know if I am good to proceed with the next steps.

File: examples/vision/adamatch.py [hunk 10]
```diff
@@ -0,0 +1,601 @@
...
+                kernel_initializer=INIT,
+                kernel_regularizer=regularizers.l2(WEIGHT_DECAY),
+                use_bias=False,
+            )(net)
+        else:
+            shortcut = net
+
+        return layers.Add()([convs, shortcut])
+
+    return f
```
Comment (fchollet) [inline examples/vision/adamatch.py:462]: Looks like a bug

File: examples/vision/adamatch.py [hunk 11]
```diff
@@ -0,0 +1,601 @@
...
+
+
+# Stacking residual Units on the same stage
+def layer(block, n_input_plane, n_output_plane, count, stride):
```
Comment (fchollet) [inline examples/vision/adamatch.py:466]: Just noticing this now, but that's *a lot* of nesting and functional indirection, which is harmful to readability. Consider doing:

```python
def block(x, n_input_plane, n_output_plane, stride):
     # One resnet block
     x = layers.BatchNormalization()(x)
     x = ...
     return x

def series_of_blocks(x, count, n_input_plane, n_output_plane, stride):
    x = block(x, n_input_plane, n_output_plane, stride=stride)
    for i in range(2, int(count + 1)):
        x = block(x, n_input_plane, n_output_plane, stride=1)
    return x

```

File: examples/vision/adamatch.py (hunk 10 above)
Comment (sayakpaul) [reply examples/vision/adamatch.py]: Resolved. Thanks.

File: examples/vision/adamatch.py (hunk 11 above)
Comment (sayakpaul) [reply examples/vision/adamatch.py]: Made it simpler for readability.

File: examples/vision/adamatch.py [hunk 12]
```diff
@@ -0,0 +1,597 @@
...
+
+The authors use a [WideResNet-28-2](https://arxiv.org/abs/1605.07146) for the dataset
+pairs we are using in this example. Most of the following code has been referred from
+[this script](https://github.com/asmith26/wide_resnets_keras/blob/master/main.py). Note
+that the following model has a scaling layer inside it that scales the pixel values to
+[0, 1].
+"""
+
+
+def wide_basic(n_input_plane, n_output_plane, stride):
```
Comment (fchollet) [inline examples/vision/adamatch.py:403]: You can further simplify the code by making this function take a tensor and return a tensor (rather than a function)

Comment (sayakpaul) [reply examples/vision/adamatch.py]: Done.

**Comment by sayakpaul** [comment]
@fchollet done. Ready for you to take the reigns.

**Review by fchollet — APPROVED** [review]
Thanks again for the great contribution! 👍
